### DISCUSSION TITLE
YAML templates are a solved problem. Why reinvent the wheel?

### DISCUSSION THREAD

**Discussion by brianthelion** [community/community#17330] [Actions]
I hate to be *that guy* -- there is A LOT to like about Github Actions -- but the UX around its YAML templates and related logic is not one of them. Specifically:

* The matrix concept is no substitute for `for` loops.
* The conditional syntax is idiomatic of absolutely nothing.

Moreover, the very idea that programming directives should somehow go *inside* YAML feels like an antipattern. YAML is for serialization; it isn't meant to program anything, much less itself. Good YAML tooling makes the distinction between the YAML and the templating directives as black-and-white as possible.

Ansible, Saltstack, et al have been working on the how-to-templatize YAML problem for a very very long time. These products are built on open source tools that have relatively large user bases. I'm sure the related communities would absolutely LOVE to move copious amounts of (jinja) code to Github Actions. Why not make it easy for them?

Why reinvent the wheel? Please just give us the option of using existing YAML templating tools.

**Comment by eric-wieser** [comment]
> The conditional syntax is idiomatic of absolutely nothing.

The conditional syntax allows the behavior of one step to be influenced by the outcome of another.
This isn't something that yaml templating can handle, because the templating would run before any of the outcomes are known.

**Reply by brianthelion** [reply]
Again, this problem has been looked at in excruciating detail by the Ansible et al. communities and AFAIK _none_ of them have fallen on the side of turning YAML into a pseudo- programming language.

Is there a trade-off between (a) keeping YAML as a pure serialization format -- eg, disallowing programming directives in the YAML -- and (b) forcing users to deal with iterative template rendering? Yes, absolutely.

But the community's experience is that the trade-offs are clearly in favor of keeping YAML immutable.
